lib/bats-core/tracing.bash

```diff
@@ -119,15 +120,25 @@ bats_trim_filename() {
 
 bats_debug_trap() {
 	# don't update the trace within library functions or we get backtraces from inside traps
-	if [[ "$1" != $BATS_ROOT/lib/* && "$1" != $BATS_ROOT/libexec/* ]]; then
-		# The last entry in the stack trace is not useful when en error occured:
-		# It is either duplicated (kinda correct) or has wrong line number (Bash < 4.4)
-		# Therefore we capture the stacktrace but use it only after the next debug
-		# trap fired.
-		# Expansion is required for empty arrays which otherwise error
-		BATS_CURRENT_STACK_TRACE=("${BATS_STACK_TRACE[@]+"${BATS_STACK_TRACE[@]}"}")
-		bats_capture_stack_trace
-	fi
+	local stack_trace_offset=
+	if [[ "$1" == $BATS_ROOT/lib/* || "$1" == $BATS_ROOT/libexec/* ]]; then
+        	if [[ "${FUNCNAME[1]}" != "run" ]]; then
+			return
+                fi
+
+                # We are inside 'run', which can throw a test failure.
+                # Make sure we log the calling function (presumably in user's
+                # .bats file) so we can emit proper line numbers in messages.
+                stack_trace_offset=1
```

Comment:
I am not sure if this is the best way to solve it. I need to think about this issue a bit more.

Reply:
Suggestions welcome!